packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -2179,7 +2177,6 @@ class StatefulElement extends ComponentElement {
         'that all the resources used by the widget are fully released.'
       );
     });
-    assert(!dirty); // See BuildableElement.unmount for why this is important.
```

Comment:
To clarify, this was the real error here. We should have removed this line when we added global key reparenting (which is when BuildableElement.unmount was removed, for the same reason). It's self-evident in retrospect but not so much all you have is the test below failing...